src/domains/chart/hooks/useAlarmBadge.js

```diff
@@ -0,0 +1,30 @@
+import { useRef } from "react"
+import { useToggle } from "react-use"
+
+export default () => {
+  const [isRendered, toggleIsRendered] = useToggle(false)
+
+  const ref = useRef(null)
+
+  const updatePosition = (alarm, g, position) => {
+    if (!alarm) {
+      if (isRendered) {
+        toggleIsRendered(false)
+      }
+      return
+    }
+
+    if (ref.current) {
+      if (!isRendered) {
+        toggleIsRendered(true)
+      }
+      const { x } = g.getArea()
+
+      ref.current.style.left = `${x}px`
+      ref.current.style.right = `calc(100% - ${position}px)`
+      ref.current.style.top = "40px"
+    }
+  }
+
+  return [isRendered, ref, updatePosition]
```

Comment:
The `updatePosition` could be defined under a useCalllback.
 
Statements like
```
    if (!alarm) {
      if (isRendered) {
        toggleIsRendered(false)
      }
      return
    }
```

Could be simplified

```
if (!alarm) return toggleIsRendered(false)
```

```
      if (!isRendered) {
        toggleIsRendered(true)
      }
      toggleIsRendered(true) // if the state is already true it will not rerender.
```

Reply:
> // if the state is already true it will not rerender.

Actually i think it will re-render, last time i checked `useState` doesn't handle opting-out from update, nor does react-use. (https://codesandbox.io/s/focused-sammet-brw2d is the official usetoggle demo if you want to experiment)


> The updatePosition could be defined under a useCalllback.

I didn't use `useCallback` because changing `updatePosition` doesn't trigger any re-renders

Reply:
This is probably related to the react version. In your example, if you change to React `16.12.0` it will not rerender.
The normal behavior for the `useState` is to prevent rerenders if we set the same value. My suggestion is to assume that it is working like this in order to write more readable code.

https://dev.to/sunflower/reactjs-if-it-is-setting-a-state-with-the-same-value-will-the-component-be-re-rendered-5g24

Reply:
of course! :) Thanks, that's how i remembered it should work, I was confused when i did codesandbox checks recently :)
done